db/memtable.cc

```diff
@@ -532,38 +533,243 @@ Status MemTable::VerifyEncodedEntry(Slice encoded,
+    *inserted_len = encoded_len;
+    return insert(table.get(), handle);
+  }
+
+  const std::lock_guard<std::mutex> l(range_del_table_lock_);
+  std::unique_ptr<MemTableIterator> iter(new MemTableIterator(
+      *this, ReadOptions(), nullptr /* arena */, true /* use_range_del_table */));
+  if (iter == nullptr) {
+    return false;
+  }
+
+  // remove the duplicates in to_insert
+  auto cmp = [](const std::tuple<Slice, Slice, SequenceNumber>& a, const std::tuple<Slice, Slice, SequenceNumber>& b) {
+    if (std::get<0>(a).data() < std::get<0>(b).data()) {
+      return -1;
+    } else if (std::get<0>(a).data() > std::get<0>(b).data()) {
+      return 1;
+    } else if (std::get<1>(a).data() < std::get<1>(b).data()) {
+      return -1;
+    } else if (std::get<1>(a).data() > std::get<1>(b).data()) {
+      return 1;
+    } else if (std::get<2>(a) < std::get<2>(b)) {
+      return -1;
+    } else if (std::get<2>(a) > std::get<2>(b)) {
+      return 1;
+    } else {
+      return 0;
+    }
+  };
+  // start, end, tombstone seqnum
+  std::set<std::tuple<Slice, Slice, SequenceNumber>, decltype(cmp)> to_insert(cmp);
+  ParsedInternalKey inkey(key, kMaxSequenceNumber, kTypeRangeDeletion);
+  std::string key_with_seq;
+  AppendInternalKey(&key_with_seq, inkey);
+  iter->SeekForPrev(Slice(key_with_seq.data(), key_with_seq.size()));
+  if (iter->Valid()) {
+    ParsedInternalKey maxkey(ExtractUserKey(iter->key()), kMaxSequenceNumber, kTypeRangeDeletion);
+    std::string k;
+    AppendInternalKey(&k, maxkey);
+    iter->Seek(Slice(k.data(), k.size()));
+  } else {
+    iter->Seek(Slice(key_with_seq.data(), key_with_seq.size()));
+  }
+  std::list<std::string> pinned_slices_;
+  auto const& user_comparator = comparator_.comparator.user_comparator();
+  for (; iter->Valid(); iter->Next()) {
+    const Slice& ikey = iter->key();
+    Slice tombstone_start_key = ExtractUserKey(ikey);
+    Slice tombstone_end_key = iter->value();
+    if (!iter->IsKeyPinned()) {
+      pinned_slices_.emplace_back(tombstone_start_key.data(),
+                                  tombstone_start_key.size());
+      tombstone_start_key = pinned_slices_.back();
+    }
+    if (!iter->IsValuePinned()) {
+      pinned_slices_.emplace_back(tombstone_end_key.data(),
+                                  tombstone_end_key.size());
+      tombstone_end_key = pinned_slices_.back();
+    }
+    SequenceNumber tombstone_seq = DecodeFixed64(tombstone_end_key.data() + tombstone_end_key.size() - 8);
+    tombstone_end_key = Slice(tombstone_end_key.data(), tombstone_end_key.size() - 8);
+    if (user_comparator->Compare(tombstone_start_key, value) >= 0) {
+      break;
+    }
+    int start_compare = user_comparator->Compare(key, tombstone_start_key);
+    if (start_compare > 0) {
+      int end_compare = user_comparator->Compare(value, tombstone_end_key);
+      if (end_compare < 0) {
+        to_insert.emplace(key, value, tombstone_seq);
+        to_insert.emplace(key, value, s);
```

Comment:
These two new fragments have the same key which our memtable [does not support](https://github.com/facebook/rocksdb/blob/a3acf2ef87e94a5877dd9561cd48f40d3075a3ee/include/rocksdb/memtablerep.h#L97). You can check the `insert(table.get(), h)` call below that `false` is returned for this fragment.

Reply:
oh, this is the exact reason why the NewTest fails, i expected b@2:c@2  and b@3:c@3 to be inserted, so when searching, b@2:c@2 will be found, i think your approach to fix NewTest works for this case, but consider this.

insert a@2: d@2
insert b@5:c@3
finding b@3, then the b@5:c@3 will be ignored, but shouldnot, because we can not use tombstone seqno to filter, we can only use fragment seqno, correct me if i am wrong.

let me think how to handle this

Reply:
I have come up with a method: encoding multiple sequence into the entry with the same start and end, how do you think about this?

By the way, this case is corresponding to the `DeleteRange(“d”, “f”)` is called at seqnum 200.` case in the doc. But with that method,  consider we have to find e@101, we have to iterate all previous entries when we found `d@200:f@200`, which is costly. what this pr's method ensures is that once you found the start key d@kMaxSequenceNumber, you only have to iterate forward, and stop once you reach a start bigger than the lookup key.

Reply:
Sorry I'm not following where does `b@5:c@3` come from. For `b@5:c@3` to be inserted, I think there should be some range tombstone inserted previously that covers 'b' at 3. 

Btw, I confused the two sequence number in the previous comment (now corrected). Fragment seqno is the seqno in key, and tombstone seqno is in value and tells us what what does this range tombstone cover.

Reply:
> encoding multiple sequence into the entry with the same start and end

Could you elaborate on the new method? 

> with the method in the doc, consider we have to find e@101, we have to iterate all previous entries when we found d@200:f@200

My method above is also a little different from the method in the doc, and it should only require one seek on the DB iter (although the seek itself on the DB iter might iterate over some entries).

 > what this pr's method ensures is that once you found the start key d@kMaxSequenceNumber, you only have to iterate forward

I thought the above test shows that this does not work, maybe I need to understand the new method first.

Reply:
yes, the full timeline maybe:
delete range a@2: d@2
delete range a@3:c@3, which will insert a@3:c@2, a@3:c@3, c@3:d@2
delete range b@5:c@5, when this encounters a@3:c@3, it will insert a@5:b@3, b@5:c@5, b@5:c@3

> Sorry I'm not following where does `b@5:c@3` come from. For `b@5:c@3` to be inserted, I think there should be some range tombstone inserted previously that covers 'b' at 3.
> 
> Btw, I confused the two sequence number in the previous comment (now corrected). Fragment seqno is the seqno in key, and tombstone seqno is in value and tells us what this range tombstone covers.

Reply:
> > encoding multiple sequence into the entry with the same start and end
> 
> Could you elaborate on the new method?
> 
> > with the method in the doc, consider we have to find e@101, we have to iterate all previous entries when we found d@200:f@200
> 
> My method above is also a little different from the method in the doc, and it should only require one seek on the DB iter (although the seek itself on the DB iter might iterate over some entries).
> 
> > what this pr's method ensures is that once you found the start key d@kMaxSequenceNumber, you only have to iterate forward
> 
> I thought the above test shows that this does not work, maybe I need to understand the new method first.

1. the new method: in FormatEntry,  when encode value, encode val_size, then encode a count, then memcpy value, then encode count sequence numbers.
2. your method (name it MA) seems not work with the case i propose in the previous reply.
3. yeah, this pr's method (name it MB) does not work in this case because the insert failure you mention earlier.

Reply:
> yes, the full timeline maybe: delete range a@2: d@2 delete range a@3:c@3, which will insert a@3:c@2, a@3:c@3, c@3:d@2 delete range b@5:c@5, when this encounters a@3:c@3, it will insert a@5:b@3, b@5:c@5, b@5:c@3

What I had in mind is for MA to handle this input in the following way:

delete range `a:d @2`
insert `a@2:d@2`

delete range `a:c @3`
overlaps with `a@2:d@2`: insert `a@3:c@3`, `c@3:d@2` (not inserting a@3:c@2 since duplicate keys)

delete range `b:c @5`
overlaps with `a@3:c@3`: insert `a@5:b@3`, `b@5:c@5` (not inserting a@5:b@3 since duplicate keys)
Note that we ignore `a@2:d@2` even thought it overlaps with `b:c@5` too since it is covered by `a@3:c@c` (same user key but with a lower sequence number).

During `Get(b, 3)`, we only look at range tombstones with fragment sequence number '<= 3` since we are reading snapshot @3, this ignores `b@5:c@5` but considers `a@3c@3` and returns NOT FOUND.

I think the confusion may be from us having different look up algorithms. As you said, MB does a `iter->SeekForPrev(key.internal_key())` to find the tombstone with the largest starting key <= input key, and then iterate forward to consider all tombstones with the same starting key. MA searches for the tombstone with the largest starting key <= input key **and** with a fragment sequence number <= snapshot sequence number.

Reply:
> the new method: in FormatEntry, when encode value, encode val_size, then encode a count, then memcpy value, then encode count sequence numbers. this way, we avoid the insert failure you mention. when searching, just decode all sequence numbers and iterate them all.

The idea looks good to me! Some thoughts to consider:
- During insertion, for each stream of tombstones with the same starting key, we only need to consider the first one and move on to the next tombstone with larger starting key. For example, consider this sequence of range deletions: `a:f@1, a:e@2, a:b@3`. During `a:b@3`, it only needs to consider `a@2:e@[2, 1]` and then can skip `a@1:f@1` and considers `e@2:f@1` next.

- In `MaxCoveringTombstoneSeqnum`, the loop may not be necessary anymore. After two seeks, we should land at the tombstone covering the look up key that gives us a full list of sequence numbers .

- I think we essentially has a representation in memtable that is much more closer to `FragmentedRangeTombstoneList`: each range tombstone has its list of sequence numbers. The construction of `FragmentedRangeTombstoneList` can be potentially optimized.

Reply:
> overlaps with `a@3:c@3`: insert `a@5:b@3`, `b@5:c@5` (not inserting a@5:b@3 since duplicate keys)
> Note that we ignore `a@2:d@2` even thought it overlaps with `b:c@5` too since it is covered by `a@3:c@c` (same user key but with a lower sequence number).

the problem seems at searching. when searching b@3, when you do SeekForPrev('b'), you will get b@5:c@5, how do you seek to `with a fragment sequence number <= snapshot sequence number.` ? you can not, because it is in the value, not ordered by skiplist. when you get b@5:c@5 you have to Prev to the first entry, or else you will miss a@3:c@3

Reply:
> > the new method: in FormatEntry, when encode value, encode val_size, then encode a count, then memcpy value, then encode count sequence numbers. this way, we avoid the insert failure you mention. when searching, just decode all sequence numbers and iterate them all.
> 
> The idea looks good to me! Some thoughts to consider:
> 
> * During insertion, for each stream of tombstones with the same starting key, we only need to consider the first one and move on to the next tombstone with larger starting key. For example, consider this sequence of range deletions: `a:f@1, a:e@2, a:b@3`. During `a:b@3`, it only needs to consider `a@2:e@[2, 1]` and then can skip `a@1:f@1` and considers `e@2:f@1` next.
> * In `MaxCoveringTombstoneSeqnum`, the loop may not be necessary anymore. After two seeks, we should land at the tombstone covering the look up key that gives us a full list of sequence numbers .
> * I think we essentially have a representation in memtable that is much more closer to `FragmentedRangeTombstoneList`: each range tombstone has its list of sequence numbers. The construction of `FragmentedRangeTombstoneList` can be potentially optimized.

I agree with you all. I will do the first two first, and maybe leave the third for future optimization.

Reply:
Sounds good! Btw I have another PR out related to range deletion optimization too #10380. It'd be interesting to run some benchmark to see the perf improvement.

Reply:
Yeah, I will run your same test for pre-pr/post-pr/your pr/scan-and-del. There is just one problem, where is the `scan-and-del` source for testing.

Reply:
by the way, can you help me see the [failed InsertWithHint test](https://app.circleci.com/pipelines/github/facebook/rocksdb/16161/workflows/cfee3990-4759-49d6-9dce-acbd37eefc13/jobs/441197), it says use an address after free, but I have no idea where the problem is.

Reply:
1. I used `--expand_range_tombstones=true` flag in db_bench for `scan-and-del` case.
2. I think you are right: reading and writing to the same shared_ptr object might not be thread-safe. I've updated the PR.